zebra-state/src/service.rs

```diff
@@ -704,7 +679,80 @@ impl Service<Request> for StateService {
                 .boxed()
             }
 
-            // TODO: add a name() method to Request, and combine all the read requests
+            // Uses pending_utxos and queued_blocks in the StateService.
+            // If the UTXO isn't in the queued blocks, runs concurrently using the ReadStateService.
+            Request::AwaitUtxo(outpoint) => {
+                metrics::counter!(
+                    "state.requests",
+                    1,
+                    "service" => "state",
+                    "type" => "await_utxo",
+                );
+
+                let timer = CodeTimer::start();
+
+                // Prepare the AwaitUtxo future from PendingUxtos.
+                let response_fut = self.pending_utxos.queue(outpoint);
+                // Only instrument `response_fut`, the ReadStateService already
+                // instruments its requests with the same span.
+                let span = Span::current();
+                let response_fut = response_fut.instrument(span).boxed();
+
+                // Check the non-finalized block queue outside the returned future,
+                // so we can access mutable state fields.
+                if let Some(utxo) = self.queued_blocks.utxo(&outpoint) {
```

Comment:
Here's one way it could return `Some`:
1. The syncer sends 20 blocks for full verification in parallel
2. The first block takes a while to verify
3. The fifth block gets verified quickly, because it doesn't depend on anything from the first four blocks, then it gets put in the queue
4. The tenth block requests a UTXO from the fifth block, which is in the queue, which returns `Some` here

Feel free to add a debug log and see how often that happens!

Reply:
Makes sense!